CONTRIBUTING.md

```diff
@@ -1,52 +1,120 @@
 
-Some things that will increase the chance that your pull request is accepted:
+When you submit a pull request, consider the following:
 
-* Use Hanami idioms
-* Include tests that fail without your code, and pass with it
-* Update the documentation, guides, etc.
+- Include tests for your change.
+- Ensure tests are passing on CI (see the “Actions” tab for each repo).
+- Add or update API documentation for your change.
+- Update user guides and/or docs to reflect your change, via a pull request to [hanami/site](https://github.com/hanami/site).
+- Add an entry to [CHANGELOG.md](./CHANGELOG.md) for your change.
+- Follow the [conventions](#conventions) outlined below, plus any other idioms you see across the codebase.
+
+If you need help with any of these, please ask! Once your pull request is ready, we’ll review your change and have a conversation with you about it.
+
+## Development setup
+
+After cloning the repo:
+
+```bash
+bundle install
+npm install
+```
+
+To run the rests:
+
+```bash
+bundle exec rake
+# or
+bundle exec rspec
+```
+
+## Conventions
+
+### Commit messages
+
+Your pull request will be squashed into a single commit when merged. Break up your pull request into individual commits if it will help with review of the code.
+
+One of your commits should contain what the message that will become the commit message for the merged pull request.
+
+[Follow these guidelines](https://developer.vonage.com/en/blog/how-to-write-a-great-git-commit-message) for your commit messages. If in doubt, don’t worry too much, we can fix things when we merge your pull request.
+
+### Changelog
+
+When you’re making a change that Hanami users should know about, please update the `CHANGELOG.md`. We follow the [keep a changelog](https://keepachangelog.com) conventions for our changelogs.
+
+New entries can go in the `## [Unreleased]` section at the top of the file. For example:
+
+```md
+## [Unreleased]
+
+### Added
+
+- Add "418 I'm a teapot" middleware for caffeine allergics (@jane_doe in #9999)
+
+## [v2.2.1] - 2024-11-12
+
+### Changed
+
+- Depend on matching minor version of hanami-cli (a version spec of `"~> 2.2.1"` instead of `"~> 2.2"`). This ensures that future bumps to the minor version of hanami-cli will not be inadvertently installed on user machines (@timriley in #1471)
+```
+
+Feel free to take inspiration from existing changelog entries for how you format your own.
+
+### API docs
+
+We use [YARD](https://yardoc.org) for our API docs.
+
+When adding or changing code, keep the API docs up to date. Only public API needs to be fully documented. Each API doc should include:
+
+- A concise, one-line description, in present tense form (see [Ruby’s API docs](https://docs.ruby-lang.org/en/master/) for examples).
+- Additional paragraphs of documentation as required.
+- Examples (via the `@example`) if hepful.
+- `@param` and `@return` tags
+- An `@api` tag
+- A `@since` tag, formatted as `@since x.x.x` if the expected release version is unknown.
```

Comment:
It's not very clear to me as a reader. Should I always add `@since x.x.x` (exactly like that) to every new method? In what situations expected release version would be known?

Reply:
Good point, @katafrakt. How about if we just phrased it as this:

> A `@since` tag, formatted as `@since x.x.x` for new entries.

I expect we'll handle this in different ways in different situations, so we can just handle it when it comes up in future PRs.

Reply:
Sounds much better IMO :+1: